EspressoSystems/jellyfish
# perf: VID ADVZ verify_share use parallelism over multiplicity

closes: #646 

### This PR does / key places to review: 
- use parallelism [here](https://github.com/EspressoSystems/jellyfish/commit/903d00773360b59ef709423f1297ba7d45315d28#diff-f9bb71cc197899825f3a7feef20ee146f64041cf3e8238288b8bed8ffd806410R566)
- new benchmark in `vid/benches/advz_multiplicity.rs` demonstrates perf improvement
- My local environment disallows me from testing whether these changes break the build when `gpu-vid` feature is enabled. Last time this issue arose @mrain had to check the build using his nvidia gpu.





### How to test this PR:
From inside the `vid/` directory run
```
RAYON_NUM_THREADS=N cargo bench --bench=advz_multiplicity --features="test-srs" 
```
On my local machine I observe significant improvement when I increase `N` from 1 to 6:

`N`=1:

```
advz_verify_payload_256KB_multiplicity/1
                        time:   [28.000 ms 29.097 ms 30.711 ms]
advz_verify_payload_256KB_multiplicity/256
                        time:   [364.81 ms 367.66 ms 371.66 ms]
```

`N`=6:

```
advz_verify_payload_256KB_multiplicity/1
                        time:   [27.515 ms 27.925 ms 28.442 ms]
                        change: [-5.5630% -2.1819% +0.6662%] (p = 0.24 > 0.05)
                        No change in performance detected.
advz_verify_payload_256KB_multiplicity/256
                        time:   [87.886 ms 89.704 ms 92.870 ms]
                        change: [-75.541% -74.950% -74.290%] (p = 0.00 < 0.05)
                        Performance has improved.
```

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [ ] Updated relevant documentation in the code
- [ ] Added relevant changelog entries to the `CHANGELOG.md` of touched crates.
